File: work/crud-nodejs/utils/imageStorage.js
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const UPLOAD_DIR = path.join(__dirname, "..", "..", "..", "uploads");

function ensureDir() {
  fs.mkdirSync(UPLOAD_DIR, { recursive: true });
}

function parseDataUrl(dataUrl) {
  const match = /^data:(image\/(\w+));base64,(.+)$/i.exec(String(dataUrl).trim());
  if (!match) return null;

  let ext = match[2].toLowerCase();
  if (ext === "jpeg") ext = "jpg";
  if (!["jpg", "png", "gif", "webp"].includes(ext)) return null;

  const buffer = Buffer.from(match[3], "base64");
  if (buffer.length > 2 * 1024 * 1024) return null;

  return { ext, buffer };
}

function saveDataUrl(dataUrl, itemId) {
  const parsed = parseDataUrl(dataUrl);
  if (!parsed) return null;

  ensureDir();
  const name = `item-${itemId}-${crypto.randomBytes(4).toString("hex")}.${parsed.ext}`;
  fs.writeFileSync(path.join(UPLOAD_DIR, name), parsed.buffer);
  return `/uploads/${name}`;
}

function deleteIfExists(imagePath) {
  if (!imagePath || !String(imagePath).startsWith("/uploads/")) return;
  const full = path.join(UPLOAD_DIR, path.basename(imagePath));
  if (fs.existsSync(full)) fs.unlinkSync(full);
}

function resolveFilePath(urlPath) {
  if (!urlPath || !String(urlPath).startsWith("/uploads/")) return null;
  const full = path.join(UPLOAD_DIR, path.basename(urlPath));
  return fs.existsSync(full) ? full : null;
}

module.exports = {
  UPLOAD_DIR,
  saveDataUrl,
  deleteIfExists,
  resolveFilePath,
};

File: work/crud-dotnet/Services/ImageStorage.cs
using System.Text.RegularExpressions;

namespace CrudDotNet.Services;

public class ImageStorage : IImageStorage
{
    private static readonly Regex DataUrlPattern = new(
        @"^data:(image/(?<type>\w+));base64,(?<data>.+)$",
        RegexOptions.Compiled | RegexOptions.IgnoreCase);

    private readonly string _uploadDir;

    public ImageStorage(IWebHostEnvironment env)
    {
        _uploadDir = Path.GetFullPath(Path.Combine(env.ContentRootPath, "..", "..", "uploads"));
        Directory.CreateDirectory(_uploadDir);
    }

    public string? SaveBase64(string dataUrl, int itemId)
    {
        var match = DataUrlPattern.Match(dataUrl.Trim());
        if (!match.Success) return null;

        var ext = match.Groups["type"].Value.ToLowerInvariant() switch
        {
            "jpeg" => "jpg",
            "jpg" => "jpg",
            "png" => "png",
            "gif" => "gif",
            "webp" => "webp",
            _ => null
        };
        if (ext is null) return null;

        var bytes = Convert.FromBase64String(match.Groups["data"].Value);
        if (bytes.Length > 2 * 1024 * 1024) return null;

        var fileName = $"item-{itemId}-{Guid.NewGuid():N[..8]}.{ext}";
        var fullPath = Path.Combine(_uploadDir, fileName);
        File.WriteAllBytes(fullPath, bytes);
        return $"/uploads/{fileName}";
    }

    public void DeleteIfExists(string? imagePath)
    {
        if (string.IsNullOrWhiteSpace(imagePath) || !imagePath.StartsWith("/uploads/"))
            return;

        var fileName = Path.GetFileName(imagePath);
        var fullPath = Path.Combine(_uploadDir, fileName);
        if (File.Exists(fullPath))
            File.Delete(fullPath);
    }
}
